p2p/http/auth/server.go

```diff
@@ -15,6 +15,29 @@ import (
 	"github.com/libp2p/go-libp2p/p2p/http/auth/internal/handshake"
 )
 
+type hmacPool struct {
+	p sync.Pool
+}
+
+func newHmacPool(key []byte) *hmacPool {
+	return &hmacPool{
+		p: sync.Pool{
+			New: func() any {
+				return hmac.New(sha256.New, key)
+			},
+		},
+	}
+}
+
+func (p *hmacPool) Get() hash.Hash {
+	return p.p.Get().(hash.Hash)
+}
+
+func (p *hmacPool) Put(h hash.Hash) {
+	h.Reset()
```

Comment:
It seems fine either way, but it looks like the example sync.Pool usage has the reset on the Get rather than the Put https://github.com/golang/go/blob/master/src/sync/example_pool_test.go#L29-L40. Similarly with this string slice pool in go-libp2p https://github.com/libp2p/go-libp2p/blob/b07e3aa6c56a4d3e8254d6689e9b32db4f0e6f61/p2p/metricshelper/pool.go#L15-L19

Reply:
Resetting in Get is better because you avoid the cost of `.Reset()` if the object is going to get GC'd anyways. In other words, you only reset things you actually use.

Thanks for highlighting this.